troposphere/haze/include/haze/event_reactor.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) Atmosphère-NX
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <haze/common.hpp>
+
+namespace haze {
+
+    class EventConsumer {
+        public:
+            virtual ~EventConsumer() = default;
+            virtual void ProcessEvent() = 0;
+    };
+
+    class EventReactor {
+        private:
+            EventConsumer *m_consumers[MAX_WAIT_OBJECTS];
+            Waiter m_waiters[MAX_WAIT_OBJECTS];
+
+            s32 m_num_wait_objects;
+            bool m_stop_requested;
+
+        public:
+            constexpr explicit EventReactor() : m_consumers(), m_waiters(), m_num_wait_objects(), m_stop_requested() { /* ... */ }
+
+            bool AddConsumer(EventConsumer *consumer, Waiter waiter);
```

Comment:
Yes. The first copy could be avoided with a pointer or reference parameter, but it would have to eventually be copied into the list in the event reactor. Maybe a move would be appropriate?